language/stdlib/modules/Testnet.move

```diff
@@ -21,5 +21,97 @@ module Testnet {
         IsTestnet{} = move_from<IsTestnet>(0xA550C18);
     }
+
+    // **************** SPECIFICATIONS ****************
+
+    /*
+       This modules is used by Genesis, LibraAccount and Unhosted as part of
+       the initialization and to check whether the transaction is on the
+       testnet.
+    */
+
+    /// # Module Specification
+
+    spec module {
+        /// Verify all functions in this module, including private ones.
+        pragma verify = true;
+
+        /// Returns the association root address.
+        define spec_root_address(): address { 0xA550C18 }
+
+        /// Returns true if the testnet has been initialized.
+        define spec_testnet_initialized(): bool {
+            exists<IsTestnet>(spec_root_address())
+        }
+    }
+
+    /// ## Management of TestNet
+
+    spec module {
+        /// **Informally:** No address has IsTestNet resource,
+        ///                 evaluates to true before initialization.
+        define spec_no_addr_has_testnet(): bool {
+            all(domain<address>(), |addr| !exists<IsTestnet>(addr))
+        }
+
+        /// **Informally:** Only the root address has an IsTestNet resource,
+        ///                 evaluates to true after initialization.
+        define spec_only_root_addr_has_testnet(): bool {
+            all(domain<address>(), |addr|
+                exists<IsTestnet>(addr)
+                    ==> addr == spec_root_address())
+        }
+    }
+
+    spec schema OnlyRootAddressHasTestNet {
+        /// Base case of the induction before the association address is
+        /// initialized with IsTestnet.
+        invariant module !spec_testnet_initialized()
+                            ==> spec_no_addr_has_testnet();
+
+        /// Inductive step after the initialization is complete.
+        invariant module spec_testnet_initialized()
+                            ==> spec_only_root_addr_has_testnet();
+    }
+
+    spec schema TestNetStaysInitialized {
+        ensures old(spec_testnet_initialized()) ==> spec_testnet_initialized();
+    }
+
+    spec module {
+        /// Apply 'OnlyRootAddressHasTestNet' to all functions.
+        apply OnlyRootAddressHasTestNet to *;
+
+        /// Apply 'TestNetStaysInitialized' to all functions except
+        /// `remove_testnet`/
+        apply TestNetStaysInitialized to * except remove_testnet;
+    }
+
+
+    /// ## Specifications for individual functions
+
+    spec module {
+        /// Returns true if IsTestNet exists at the address of the account.
+        define spec_exists_testnet(account: signer): bool {
+            exists<IsTestnet>(Signer::get_address(account))
+        }
+    }
+
+    spec fun initialize {
+        aborts_if Signer::get_address(account) != spec_root_address();
+        aborts_if spec_exists_testnet(account);
+        ensures spec_exists_testnet(account);
+    }
+
+    spec fun is_testnet {
+        aborts_if false; // why does ading aborts_if true give me "postcondition might not hold"?
```

Comment:
Please add a TODO if you add comments like this, so we can find such places and remove them from the final source.

I guess that is the question we discussed in the standup. Could you please open an issue for this?